pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -444,6 +445,268 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
+        // The max weight actually depends on the number of other tokens, but this is handled below
+        // by ensuring the final weights are all >= minimum
+        _require(normalizedWeight < FixedPoint.ONE, Errors.MAX_WEIGHT);
+        _require(_getTotalTokens() < _getMaxTokens(), Errors.MAX_TOKENS);
+
+        // Do not allow adding tokens if there is an ongoing or pending gradual weight change
+        uint256 currentTime = block.timestamp;
+        bytes32 poolState = _getMiscData();
+
+        if (currentTime < poolState.decodeUint32(_WEIGHT_START_TIME_OFFSET)) {
+            _revert(Errors.CHANGE_TOKENS_PENDING_WEIGHT_CHANGE);
+        } else if (currentTime < poolState.decodeUint32(_WEIGHT_END_TIME_OFFSET)) {
+            _revert(Errors.CHANGE_TOKENS_DURING_WEIGHT_CHANGE);
+        }
+
+        uint256 weightSumBeforeAdd = _denormWeightSum;
+
+        // Calculate the weightSum after the add
+        // Consider an 80/20 pool:
+        // |----0.8----|-0.2-|
+        // 0                 x = 1.0 (rightmost point at the "end" of the number line)
+        //
+        // Now add a new token with a weight of 60%
+        // |----0.8----|-0.2-|---0.6y---|
+        // 0                 x          y = the new weightSum
+        //
+        // By definition, since we interpret the new token weight as the desired final normalized weight,
+        // the new "length" is the old length, plus 60% of the total new length, y
+        // y = 0.6y + x
+        // (1 - 0.6)y = x
+        // y = x / (1 - 0.6); since x = 1, y = 1/0.4 = 2.5
+        //
+        // The denormalized weight of the new token, 0.6y, is then 0.6(2.5) = 1.5
+        // Since the denorm weights of the original tokens stay the same, the final state is:
+        // |----0.8----|-0.2-|---1.5---|  denormalized weights (as stored)
+        //    0.8/2.5  0.2/2.5 1.5/2.5
+        //      32%      8%      60%      normalized weights (as calculated: W[i]/weightSum)
+        //
+        // The added token is 60%, as desired, and the original 80/20 weights are scaled down
+        // proportionately to 32/8, to fit within the remaining 40%
+        //
+        //uint256 weightSumMultiplier = FixedPoint.ONE.divDown(FixedPoint.ONE - normalizedWeight);
+        weightSumAfterAdd = weightSumBeforeAdd.mulUp(FixedPoint.ONE.divDown(FixedPoint.ONE - normalizedWeight));
+
+        // Now we need to make sure the other token weights don't get scaled down below the minimum
+        // normalized weight[i] = denormalized weight[i] / weightSumAfterAdd
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        for (uint256 i = 0; i < tokens.length; i++) {
+            _require(
+                _getTokenData(tokens[i]).decodeUint64(_END_DENORM_WEIGHT_OFFSET).uncompress64(_MAX_DENORM_WEIGHT).divUp(
+                    weightSumAfterAdd
+                ) >= WeightedMath._MIN_WEIGHT,
+                Errors.MIN_WEIGHT
+            );
+        }
```

Comment:
Processing _all_ tokens seems unnecessarily wasteful. All we need to do is check that no token goes below the minimum weight, but we can do that much more efficiently by keeping track of the smallest weight, and comparing against that.

Weights only change in add token, remove token, and gradual weight change. In a gradual weight change we process all tokens, so we can easily find the smallest weight and store it. We only care about the smallest end weight, since add/remove token is disabled until the weight change finishes. Remove token also changes weights by increasing them by a known factor, and add token similarly decreases them by a factor: we can apply this same factor to the smallest weight.

Reply:
I don't think this makes sense from a gas perspective.

With this method we're performing `numTokens` warm SLOADs. Best case scenario we're saving `(2 * numTokens - 1) * 100` gas (assuming we're packing this cached weight so the SLOAD is warm).

The flipside is that we need to update the cache on gradual weight updates which change the weight of the lowest weighted token, which costs 5000 gas (I don't see a way to pack things which allows us to do this SSTORE for free).

Weight updates are going to be way more common than adding new tokens to the pool and unless you have 20+ tokens in your pool then you'll wipe out the savings by changing the weights just once.

Reply:
Plus less code is good.